### PR TITLE
Test Network Functionalities and Fix Related Bugs

### PR DESCRIPTION
**Pull request by khoaguin**
## Description
Testing Syft's networking functionalities, including
- Launching a gateway node
- Connecting to a gateway from domain nodes
- Listing gateways with `sy.gateways` and domains with `sy.domains`
- Searching datasets via the gateway with `sy.search` using `ThreadPoolExecutor`
- Gateway client obtains a `ProxyClient` to the domain client 
- Add an environment variable which lets us inject a json string for the network registry which gets used instead of the url on startup for testing purposes

And also fix related bugs (https://github.com/OpenMined/Heartbeat/issues/1140) and some improvements (https://github.com/OpenMined/Heartbeat/issues/1139)

## How has this been tested?
Added some integration tests in `tests/integration/network/gateway_test.py`

### CODE REVIEW

**Comment by khoaguin** [comment]
Got this `TypeHintWarning` that needs to be fixed when login
image

File: packages/syft/src/syft/types/syft_object.py [hunk 1]
```diff
@@ -587,11 +587,11 @@ def _syft_set_validate_private_attrs_(self, **kwargs: Any) -> None:
             if value is not PydanticUndefined:
                 if var_annotation is not None:
                     # Otherwise validate value against the variable annotation
-                    check_type(
-                        value=value,
-                        expected_type=var_annotation,
-                        forward_ref_policy=ForwardRefPolicy.IGNORE,
-                    )
+                    with std_stream_capture():
```
Comment (khoaguin) [inline packages/syft/src/syft/types/syft_object.py:590]: Somehow using this makes all the stack notebook tests time out

**Comment by khoaguin** [comment]
After launching a gateway node and a domain node, connect the domain node to the gateway node, we have

Screenshot from 2024-03-14 14-40-11

**Comment by khoaguin** [comment]
Error while trying to launch and login web-server gateway nodes

image

image

image

**Comment by khoaguin** [comment]
https://files.slack.com/files-pri/T6963A864-F06Q2RJRRDF/image.png

**Comment by khoaguin** [comment]
Progress: fix bugs in https://github.com/OpenMined/Heartbeat/issues/1140

image
image

Note: now `DatasetService.search`'s roles is `GUEST_ROLE_LEVEL`

**Review by shubham3121 — APPROVED** [review]

**Comment by shubham3121** [comment]
Nice work @khoaguin with the fixes !!

**Pull request merged**